### PR TITLE
New module testing framework

### PR DESCRIPTION
**Pull request by shinae-woo**
Module testing framework including improvement suggestions in #625.

The major changes are
- Adopting python unittest module
- Checking module scheduling completion by checking traffic class statistics (No more sleep and blocking system call, shorter test completion time) 

I did not heavily fix each module test, but more focusing on porting existing tests into the new framework. It would be great if the tests are gradually improved as modules are updated.

Reporting two issues
1. To get the benefit of syntactic sugar, the tests are still called by `bessctl`. Each module test is called as `exec()`, we can not directly get the return value from unittest.  In the case of test failure, we need to forcibly exit the tests by returning a non-zero value.  This can be further improved by moving sugar semantics (+ more functionality) from `bessctl` to 'pybess' so that native unittest.py can be directly used. Separate PR required.

2. All tests in the same file share the same execution environments (by `exec()`). This may become problematic when we use sugared statements which automatically add global variables. When we use `src::Source()` multiple times for different test functions in a single file,  the second src::Source() call will fail due to conflict module names, since `src` variable is global, not local.

### CODE REVIEW

File: bessctl/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            raise AssertionError('"%s" != "%s"',
+                                 pkt_str(pkt1), pkt_str(pkt2))
+
+    def assertNotSamePackets(self, pkt1, pkt2):
+        if pkt_str(pkt1) == pkt_str(pkt2):
+            raise AssertionError('"%s" == "%s"',
+                                 pkt_str(pkt1), pkt_str(pkt2))
+
+    def assertBessAlive(self):
+        self.bess = BESS()
```
Comment (sangjinhan) [inline bessctl/test_utils.py]: Recreating (and replacing) a `BESS` instance may be somewhat disruptive. We can either:
1. using existing `self.bess`, we check with a dummy RPC (e.g., `GetVersion`), or
2. we create a new BESS instance, but we do not overwrite self.bess with it.

File: bessctl/conf/testing/run_module_tests.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+def main():
+    any_failure = 0
+
+    try:
+        run_cmd('%s daemon start' % bessctl)
+    except CommandError:
+        raise Exception('bess daemon could not start')
+
+    for file_name in glob.glob(os.path.join(test_dir, "*.py")):
```
Comment (nemethf) [inline bessctl/conf/testing/run_module_tests.py]: Wouldn't this mean that the feature of testing just a single module disappears?  I've found that very useful.

Comment (shinae-woo) [reply bessctl/conf/testing/run_module_tests.py]: I propose a new one! Please check a `assertBessAlive()` in `test_util.py` and see sample code in module test code.

Comment (sangjinhan) [reply bessctl/conf/testing/run_module_tests.py]: Maybe I misunderstand, but I guess he meant the configurable `TEST_PATH` and `TEST_NAME` environment variable  here, not about testing "a single module" vs "a chain of modules"

Comment (shinae-woo) [reply bessctl/conf/testing/run_module_tests.py]: Oh, sorry. it was 'my' misunderstand. @nemethf, you can simply run a single test within `bessctl`. For example, simply do `run testing/xxx.py`, then the test results will show up on the shell.

**Comment by shinae-woo** [comment]
@sangjinhan, Of course, I will do it. I see that way is much better for maintenance.

**Comment by shinae-woo** [comment]
Reminder for myself, in an offline discussion, @chris3torek pointed out that the current `run_module()` is not fully equivalent with existing  `CRASH_TEST_INPUTS`  as now we use random packets, but previously they use predefined (I would say `hard-coded) set of packets. Using random packets are more appropriate for the most case for checking crash rather than correct functionality, but I also quite agree that it would be better if there is a way that users can specify a set of packet template to run. I will extend `run_module()` interface accordingly.

File: bessctl/test_utils.py (hunk 1 above)
Comment (shinae-woo) [reply bessctl/test_utils.py]: Good point.

File: bessctl/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    if not pkt:
+        return str('').encode('HEX')
+    else:
+        return str(pkt).encode('HEX')
+
+
+class BessModuleAssertion():
+
+    def assertSamePackets(self, pkt1, pkt2):
+        if pkt_str(pkt1) != pkt_str(pkt2):
```
[the remark this answers is not in the document]
Comment (shinae-woo) [reply bessctl/test_utils.py]: `assertEquals()` are not available unless inheriting `unittest.Test` as you said. But they show the same behavior (raising same `AssertionError`), and we can also customize assert error messages per error type in this way :)

File: bessctl/conf/testing/run_module_tests.py (hunk 2 above)
Comment (nemethf) [reply bessctl/conf/testing/run_module_tests.py]: Oh, that's great!  (And sorry for the noise.)

**Comment by sangjinhan** [comment]
Merged, thanks!
c817c23ed5406b8cbfdf17e95ed206af
